### PR TITLE
Fix qpy for controlled `UnitaryGate`

### LINKED ISSUE #10802: Serialization fails for controlled unitary gate
**Describe the bug**
I get the error: `AttributeError: 'NoneType' object has no attribute 'clbits` and it seems that the `Qpy to QASM conversion failed: unable to parse quantum circuit` for a circuit containing a controlled unitary gate.

**Steps to reproduce**
```
from qiskit.extensions import UnitaryGate
from qiskit.quantum_info import random_unitary, SparsePauliOp
from qiskit import QuantumCircuit, QuantumRegister
from qiskit_ibm_runtime import QiskitRuntimeService, Options, Estimator, Session

UU = random_unitary(4, seed=12345)
qr = QuantumRegister(3)
qc = QuantumCircuit(qr)
qc.h(0)
qc.append(UnitaryGate(UU).control(1), qr)

service = QiskitRuntimeService()
backend = service.get_backend('ibmq_qasm_simulator')

options = Options(optimization_level=2, resilience_level=0)
estimator = Estimator(backend=backend, options=options)

XXexp = estimator.run(qc, observables=SparsePauliOp('IIX')).result().values[0]
```

**Expected behavior**
It should work

**Suggested solutions**

**Additional Information**

Transpiling before works
`transpile(qc, basis_gates = backend.configuration().basis_gates)`

- **qiskit-ibm-runtime version**: 0.12.0
- **Python version**: 3.8.16
- **Operating system**: MacOs

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
Fixes #10802. QPY was failing to deserialize the `NUMPY_OBJ` from the unitary gate definition because it was looking for the `clbits` of a circuit that, in that scope, was `None`. 

### Details and comments
I am not too happy about having to decompose the circuit so many times for the test. I get that after #10758, the uuid is appended to the name of the gate, which makes the circuits not equal:

```
>>> qc.data
[CircuitInstruction(operation=Instruction(name='c-unitary', num_qubits=2, num_clbits=0, params=[array([[0.+0.j, 1.+0.j],
       [1.+0.j, 0.+0.j]])]), qubits=(Qubit(QuantumRegister(2, 'q'), 0), Qubit(QuantumRegister(2, 'q'), 1)), clbits=())]
>>> qc2.data
[CircuitInstruction(operation=Instruction(name='c-unitary_781f2d2b-cd9a-4d8e-b38a-59d141b82ebc', num_qubits=2, num_clbits=0, params=[array([[0.+0.j, 1.+0.j],
       [1.+0.j, 0.+0.j]])]), qubits=(Qubit(QuantumRegister(2, 'q'), 0), Qubit(QuantumRegister(2, 'q'), 1)), clbits=())]
```

But I'm not sure why, for example, decomposing once fails:

```
>>> qc.decompose(reps=1).data
[CircuitInstruction(operation=Instruction(name='circuit-121_dg', num_qubits=2, num_clbits=0, params=[]), qubits=(Qubit(QuantumRegister(2, 'q'), 0), Qubit(QuantumRegister(2, 'q'), 1)), clbits=())]
>>> qc2.decompose(reps=1).data
[CircuitInstruction(operation=Instruction(name='circuit-121_dg', num_qubits=2, num_clbits=0, params=[]), qubits=(Qubit(QuantumRegister(2, 'q'), 0), Qubit(QuantumRegister(2, 'q'), 1)), clbits=())]
```

### CODE REVIEW

File: test/python/circuit/test_circuit_load_from_qpy.py [hunk 1]
```diff
@@ -169,6 +169,20 @@ def test_unitary_gate(self):
...
+        """Test that numpy array parameters are correctly serialized
+        in controlled unitary gate."""
+        qc = QuantumCircuit(2)
+        unitary = np.array([[0, 1], [1, 0]])
+        gate = UnitaryGate(unitary)
+        qc.append(gate.control(1), [0, 1])
+        qpy_file = io.BytesIO()
+        dump(qc, qpy_file)
+        qpy_file.seek(0)
+        new_circ = load(qpy_file)[0]
```
Comment (mtreinish) [inline test/python/circuit/test_circuit_load_from_qpy.py]: Just to ensure the buffer gets closed:

```suggestion
        with io.BytesIO() as qpy_file:
            dump(qc, qpy_file)
            qpy_file.seek(0)
            new_circ = load(qpy_file)[0]
```

**Review by mtreinish — COMMENTED** [review]
The fix LGTM, but there were some merge conflicts that need to be resolved. This also might have a conflict with #10898 but I assume this will merge before that, so I'll fix it in #10898 when this merges.

**Review by mtreinish — DISMISSED** [review]
Thanks for the quick update.

(oh there is still a merge conflict)

File: qiskit/qpy/binary_io/circuits.py [hunk 2]
```diff
@@ -577,12 +577,21 @@ def _write_instruction(file_obj, instruction, custom_operations, index_map, use_
         or gate_class_name == "Instruction"
         or isinstance(instruction.operation, library.BlueprintCircuit)
     ):
-        if instruction.operation.name not in custom_operations:
-            custom_operations[instruction.operation.name] = instruction.operation
-            custom_operations_list.append(instruction.operation.name)
         gate_class_name = instruction.operation.name
+        # ucr*-dg gates can have different numbers of parameters,
+        # the uuid is appended to avoid storing a single definition
+        # in circuits with multiple ucr*-dg gates.
```
Comment (ElePT) [inline qiskit/qpy/binary_io/circuits.py]:
```suggestion
        # in circuits with multiple ucr*_dg gates.
```

File: qiskit/qpy/binary_io/circuits.py [hunk 3]
```diff
@@ -577,12 +577,21 @@ def _write_instruction(file_obj, instruction, custom_operations, index_map, use_
...
+            gate_class_name += "_" + str(uuid.uuid4())
+
+        if gate_class_name not in custom_operations:
+            custom_operations[gate_class_name] = instruction.operation
+            custom_operations_list.append(gate_class_name)
 
     elif gate_class_name == "ControlledGate":
+        # controlled gates can have the same name but different parameter
+        # values, the uuid is appended to avoid storing a single definition
+        # in circuits with multiple ucr*-dg gates.
```
Comment (ElePT) [inline qiskit/qpy/binary_io/circuits.py]:
```suggestion
        # in circuits with multiple controlled gates.
```

**Review by mtreinish — DISMISSED** [review]
This LGTM, thanks for the update, but I'll hold off on the merge queue because something looks weird with the last rebase, it pulled in a bunch of unrelated commits to your branch.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**